### PR TITLE
Support `GenericModel` and `AbstractModel`

### LINKED ISSUE #85: Generalize to `AbstractModel`s
To make DisjunctiveProgramming compatible with JuMP and InfiniteOpt. We'll need to generalize to handle `AbstractModel` and `AbstractVariableRef`. The trickiest part will be figuring out what to do with `LogicalVariableRef`s. This should be developed in tandem with whatever needs to happen on the InfiniteOpt side of things.

### PR DESCRIPTION
**Pull request by pulsipher**
This closes #85. This adds new constructors for `GDPModel` to support `JuMP.AbstractModel`s, `JuMP.GenericModel`s, and other types of variable/constraint references. This paves the way for this package to work with InfiniteOpt.

The major changes include:
- Parameterizing many of the types by the model, variable, and/or constraint type
- Storing variable bounds directly in `GDPData` since this is commonly used by several reformulation methods
- Creating binary variables immediately when logical variables are added and adding `binary_variable` to access them (this is needed to embed logical variables in algebraic constraints which InfiniteOpt needs)
- Cleaning up the docs a little more
- Changing the `method` keyword in `optimize!` to `gdp_method` for better clarity --> needed for InfiniteOpt

### CODE REVIEW

**Comment by pulsipher** [comment]
For reference, the following will be needed as an extension to enable the use of InfiniteOpt. Currently, this relies on the master branch of InfiniteOpt and can be included in another PR once InfiniteOpt makes its next release.
```julia
import DisjunctiveProgramming as DP
import InfiniteOpt, JuMP

function InfiniteGDPModel(args...; kwargs...)
    return DP.GDPModel{
        InfiniteOpt.InfiniteModel, 
        InfiniteOpt.GeneralVariableRef, 
        InfiniteOpt.InfOptConstraintRef
        }(args...; kwargs...)
end

InfiniteLogical(prefs...) = DP.Logical(InfiniteOpt.Infinite(prefs...))

function DP.requires_disaggregation(vref::InfiniteOpt.GeneralVariableRef)
    if vref.index_type <: InfiniteOpt.InfOptParameter
        return false
    else
        return true
    end
end

function DP.make_disaggregated_variable(
    model::InfiniteOpt.InfiniteModel,
    vref::InfiniteOpt.GeneralVariableRef,
    name,
    lb,
    ub
    )
    prefs = InfiniteOpt.parameter_refs(vref)
    if isempty(prefs)
        return JuMP.@variable(model, base_name = name, lower_bound = lb, upper_bound = ub, 
                              variable_type = InfiniteOpt.Infinite(prefs...))
    else
        return JuMP.@variable(model, base_name = name, lower_bound = lb, upper_bound = ub)
    end
end

function JuMP.add_constraint(
    model::InfiniteOpt.InfiniteModel,
    c::JuMP.VectorConstraint{F, S},
    name::String = ""
    ) where {F, S <: DP.AbstractCardinalitySet}
    return DP._add_cardinality_constraint(model, c, name)
end
function JuMP.add_constraint(
    model::M,
    c::JuMP.ScalarConstraint{DP._LogicalExpr{M}, S},
    name::String = ""
    ) where {S, M <: InfiniteOpt.InfiniteModel}
   return DP._add_logical_constraint(model, c, name)
end
function JuMP.add_constraint(
    model::M,
    c::JuMP.ScalarConstraint{DP.LogicalVariableRef{M}, S},
    name::String = ""
    ) where {M <: InfiniteOpt.InfiniteModel, S}
    error("Cannot define constraint on single logical variable, use `fix` instead.")
end
function JuMP.add_constraint(
    model::M,
    c::JuMP.ScalarConstraint{JuMP.GenericAffExpr{C, DP.LogicalVariableRef{M}}, S},
    name::String = ""
    ) where {M <: InfiniteOpt.InfiniteModel, S, C}
    error("Cannot add, subtract, or multiply with logical variables.")
end
function JuMP.add_constraint(
    model::M,
    c::JuMP.ScalarConstraint{JuMP.GenericQuadExpr{C, DP.LogicalVariableRef{M}}, S},
    name::String = ""
    ) where {M <: InfiniteOpt.InfiniteModel, S, C}
    error("Cannot add, subtract, or multiply with logical variables.")
end
```
With the above, then we can write infinite GDP models and reformulate them before discretizing which improves performance:
```julia
using InfiniteOpt, DisjunctiveProgramming, HiGHS

# Initialize the model
model = InfiniteGDPModel(HiGHS.Optimizer)

# Create the infinite variables
I = 1:4
@infinite_parameter(model, t ∈ [0, 1], num_supports = 100)
@variable(model, 0 <= g[I] <= 10, Infinite(t))

# Add the disjunctions and their indicator variables
@variable(model, G[I, 1:2], InfiniteLogical(t))
@constraint(model, [i ∈ I, j ∈ 1:2], 0 <= g[i], Disjunct(G[i, 1]))
@constraint(model, [i ∈ I, j ∈ 1:2], g[i] <= 0, Disjunct(G[i, 2]))
@disjunction(model, [i ∈ I], G[i, :])

# Add the logical propositions
@variable(model, W, InfiniteLogical(t))
@constraint(model, G[1, 1] ∨ G[2, 1] ∧ G[3, 1] == W := true)
@constraint(model, 𝔼(binary_variable(W), t) >= 0.95)

# Reformulate and solve 
optimize!(model, gdp_method = Hull())
```

File: src/variables.jl [hunk 1]
```diff
@@ -43,15 +85,21 @@ Extend `JuMP.add_variable` for [`LogicalVariable`](@ref)s. This
...
     )
     is_gdp_model(model) || error("Can only add logical variables to `GDPModel`s.")
-    data = LogicalVariableData(v, name)
+    # add the logical variable
+    data = LogicalVariableData(_get_variable(v), name)
     idx = _MOIUC.add_item(_logical_variables(model), data)
+    lvref = LogicalVariableRef(model, idx)
     _set_ready_to_optimize(model, false)
-    return LogicalVariableRef(model, idx)
+    # add the assocviated binary variables
```
Comment (hdavid16) [inline src/variables.jl:98]: typo: assocviated

File: test/constraints/selector.jl [hunk 2]
```diff
@@ -174,6 +174,18 @@ function test_nested_atmost_reformulation()
         DP._indicator_to_binary(model)[y[3]]
 end
 
+function test_extension_variables()
+    model = GDPModel{MyModel, MyVarRef, MyConRef}()
+    @variable(model, y[1:2], Logical(MyVar), start = true)
+    cref = LogicalConstraintRef(model, LogicalConstraintIndex(1))
+    @test @constraint(model, y in AtLeast(1)) == cref
+    @test reformulate_model(model, DummyReformulation()) isa Nothing
```
Comment (hdavid16) [inline test/constraints/selector.jl:182]: With having BigM as the default, you can remove the DummyReformulation here unless we want to keep it.

Comment (pulsipher) [reply test/constraints/selector.jl:182]: We can keep it as an extra Reformulation method to test extension methods

**Comment by pulsipher** [comment]
> looks good. Thanks! I added default a reformulation method (BigM) to `master`. Can you rebase/resolve any conflicts?

Thanks for reviewing, it should be good to go, assuming all the tests pass.

**Review by hdavid16 — APPROVED** [review]

**Pull request merged**